src/accelerate/accelerator.py

```diff
@@ -2120,8 +2122,29 @@ def gather_for_metrics(self, tensor):
         9
         ```
         """
-        tensor = self.gather(tensor)
 
+        def check_input(input_data):
+            if isinstance(input_data, (list, tuple)):
+                return all(is_torch_tensor(item) for item in input_data)
+            return is_torch_tensor(input_data)
+
+        if not check_input(input_data):
+            try:
+                obj = gather_object(input_data)
+                # deduplicate objects within the last batch
+                if self.gradient_state.end_of_dataloader and self.gradient_state.remainder > 0:
+                    deduplicated_objects = []
+                    seen_objects = set()
+                    for item in obj:
+                        if item not in seen_objects:
+                            deduplicated_objects.append(item)
+                            seen_objects.add(item)
+                return deduplicated_objects
+
+            except Exception:
+                return obj
+
+        tensor = self.gather(input_data)
```

Comment:
I think we can do a bit better here:

```suggestion
        try:
            recursively_apply(lambda x: x, input_data, error_on_other_type=True)
            all_tensors = True
        except TypeError:
            all_tensors = False

        if not all_tensors:
            data= gather_object(input_data)

        data= self.gather(input_data)
```

From here, correct me if I'm wrong but the following `try/except` should then be able to filter out the deuplicates.

Note that `gather_for_metrics` should only remove the duplicated items at the end, not *all duplicates in the batch*, hence why we need this indexing. Does this make sense?